### PR TITLE
Add windows JNI support

### LINKED ISSUE #32516: Java bindings for Windows
## 🚀 Feature
It would be cool if the Java bindings worked on Windows.  They don't right now.  If you are interested in this feature, please comment on this issue with information on how/why you want to use it.  We will use this information to prioritize the work.

cc @peterjc123

### PR DESCRIPTION
**Pull request by erip**
Fixes #32516

### CODE REVIEW

**Comment by erip** [comment]
I'm not positive if there's something special I need to do to ensure a JDK is installed on the CI runners, but the usual suspects on Windows (`C:\Program Files\Java`, `C:\Program Files\AdoptOpenJDK`) aren't resolving to `JAVA_HOME`s. 😬

File: .circleci/scripts/binary_populate_env.sh [hunk 1]
```diff
@@ -100,6 +100,11 @@ if [[ "$PACKAGE_TYPE" == libtorch ]]; then
   POSSIBLE_JAVA_HOMES+=(/usr/local)
   POSSIBLE_JAVA_HOMES+=(/usr/lib/jvm/java-8-openjdk-amd64)
   POSSIBLE_JAVA_HOMES+=(/Library/Java/JavaVirtualMachines/*.jdk/Contents/Home)
+  # Using 8.3 filenames is a bit annoying...
+  POSSIBLE_JAVA_HOMES+=(/c/Progra~1/Java/jdk*)
```
Comment (peterjc123) [inline .circleci/scripts/binary_populate_env.sh:104]: Why do we need to use 8.3 filenames here? Does the following one work?
```
POSSIBLE_JAVA_HOMES+=("/c/Program Files/Java/jdk*")
POSSIBLE_JAVA_HOMES+=("/c/Program Files/OpenJDK/jdk*")
POSSIBLE_JAVA_HOMES+=("/c/Program Files (x86)/Java/jdk*")
POSSIBLE_JAVA_HOMES+=("/c/Program Files (x86)/OpenJDK/jdk*")
```

Comment (erip) [reply .circleci/scripts/binary_populate_env.sh:104]: I had tried those, but it wasn't working locally when debugging. It seems like it's not getting picked up in any case.

**Comment by mszhanyi** [comment]
Fix #44363 too

**Review by mszhanyi — COMMENTED** [review]
sorry, ignore it

**Review by mszhanyi — COMMENTED** [review]
Fixes #44363

**Review by mszhanyi — APPROVED** [review]

**Comment by erip** [comment]
This is not quite ready, so I'd remove approval @mszhanyi. The fbjni folks have also been responding, so this may be moving a bit more soon. 🙌

**Comment by mszhanyi** [comment]
> I'm not positive if there's something special I need to do to ensure a JDK is installed on the CI runners, but the usual suspects on Windows (`C:\Program Files\Java`, `C:\Program Files\AdoptOpenJDK`) aren't resolving to `JAVA_HOME`s. 😬
No problem. I just don't find how to remove me from the reviewer list.

**Comment by erip** [comment]
Lingering TODOs: 

- conditionally link optional libraries on Windows.
- Find out why libtorch build in failing on Linux. 🤔

**Comment by erip** [comment]
I found why cxx11 was failing. As of [a85423b](https://github.com/facebookincubator/fbjni/commit/a85423bd11992aec3dec44ed9c87a518c598ac91#diff-af3b638bc2a3e6c650974192a53c7291), fbjni is C++14-only.

**Comment by erip** [comment]
cc @dreiss do these fbjni errors look familiar to you? They they all look related and [this](https://github.com/facebookincubator/fbjni/commits/master/cxx/fbjni/detail/SimpleFixedString.h) is the history of `SimpleFixedString.h` - not sure why it only fails in these limited cases...

**Comment by erip** [comment]
@mszhanyi I don't imagine this PR is going to change dramatically by fixing the remaining two build errors, so if you'd like to review it I think this PR is quite close.

**Comment by erip** [comment]
Waiting for [this PR](https://github.com/facebookincubator/fbjni/pull/40) to be merged.

**Comment by erip** [comment]
I think these errors are a fixture of flaky builds and upstream breakages. I'll try to rebuild a bit later.

**Comment by erip** [comment]
d2248cd is a temporary commit while we wait for the commits to land in fbjni.

**Review by peterjc123 — APPROVED** [review]
These changes LGTM.

**Review by mszhanyi — APPROVED** [review]

**Comment by erip** [comment]
@peterjc123 @mszhanyi I think this is good once CI passes. WDYT?

File: android/pytorch_android/CMakeLists.txt [hunk 2]
```diff
@@ -122,19 +122,37 @@ if(ANDROID_ABI)
...
       c10
-      nnpack
-      XNNPACK
-      pytorch_qnnpack
-      pthreadpool
       cpuinfo
       clog
   )
+
+  if(nnpack_FOUND)
```
Comment (t-vi) [inline android/pytorch_android/CMakeLists.txt:140]: Here (and in the other libs), you're making the libraries optional on non-windows platforms as well.
The main CMakeLists.txt seems to use a USE_NNPACK option instead. If this difference intentional?

Comment (erip) [reply android/pytorch_android/CMakeLists.txt:140]: Not intentional - just a lack of experience with cmake. 😄  It seems like I should just be able to replace the `find_package` and `if(*_FOUND)` with appropriate `if(USE_*)` checks - I can fix that up.

Comment (t-vi) [reply android/pytorch_android/CMakeLists.txt:140]: I'll liberally admit that I wouldn't know if it is a problem or not, so an alternative strategy to addressing my drive-by comment would be to see if someone who actually knows something (@dreiss or @ljk53 or one of the other cool people) have something to say.

Comment (erip) [reply android/pytorch_android/CMakeLists.txt:140]: I will commit the change with the option to revert as needed. 😄

Comment (erip) [reply android/pytorch_android/CMakeLists.txt]: See e62256e2289

Comment (erip) [reply android/pytorch_android/CMakeLists.txt]: The failure looks like a one-off docker thing - it _seems_ unrelated to the fix. I rebased and that may have caused something funny to happen.

**Review by peterjc123 — APPROVED** [review]
Again, this LGTM.

**Comment by peterjc123** [comment]
Please merge this pr. @ezyang @malfet

**Comment by erip** [comment]
These errors seem to be related to git funniness - seems to be a fluke and probably should be OK to merge.

**Pull request closed without merging**